File: src/Core/CoreLib/Collections/Generic/Queue.cs
// Queue.cs
// Script#/Libraries/CoreLib
// This source code is subject to terms and conditions of the Apache License, Version 2.0.
//

using System.Runtime.CompilerServices;

namespace System.Collections.Generic {

    [ScriptImport]
    [ScriptName("Queue")]
    public sealed class Queue<T> {

        [ScriptField]
        public int Count {
            get {
                return 0;
            }
        }

        public void Clear() {
        }

        public bool Contains(T item) {
            return false;
        }

        public T Dequeue() {
            return default(T);
        }

        public void Enqueue(T item) {
        }

        public T Peek() {
            return default(T);
        }
    }
}

File: src/DSharp.Script/src/System/Collections/Queue.js
function Queue() {
    this.count = 0;
    this._items = [];
    this._offset = 0;
}
function _cleanQueue(q) {
    q._items = q._items.slice(q._offset);
    q._offset = 0;
}
var Queue$ = {
    clear: function () {
        this._items.length = 0;
        this._offset = 0;
        this.count = 0;
    },
    contains: function (item) {
        for (var i = this._offset, length = this._items.length; i <= length; i++) {
            if (this._items[i] === item) {
                return true;
            }
        }
        return false;
    },
    dequeue: function () {
        if (this.count) {
            var item = this._items[this._offset];
            if (++this._offset * 2 >= this._items.length) {
                _cleanQueue(this);
            }
            this.count--;
            return item;
        }
        return undefined;
    },
    enqueue: function (item) {
        this._items.push(item);
        this.count++;
    },
    getEnumerator: function () {
        if (this._offset != 0) {
            _cleanQueue(this);
        }
        return new Enumerator(this._items);
    },
    peek: function () {
        return this._items.length
            ? this._items[this._offset]
            : undefined;
    }
};